### PR TITLE
support setting pipeline env vars from the config

### PR DESCRIPTION
**Pull request by dustymabe**
See individual commit messages.

Relies on https://github.com/coreos/coreos-ci-lib/pull/157

### CODE REVIEW

**Comment by dustymabe** [comment]
marking as draft.. i'd like to merge this myself so I can be around afterwards to clean up any fallout. I did do some minimalish testing in stage.

**Comment by dustymabe** [comment]
hmm. not sure what that CI failure is trying to get me to do? rename to `makeCosaRemoteSession()`?

**Comment by jlebon** [comment]
> hmm. not sure what that CI failure is trying to get me to do? rename to `makeCosaRemoteSession()`?

Yeah, that's annoying.

I found https://docs.codacy.com/faq/code-analysis/how-to-skip-an-analysis/ which seems like a way to work around this.

**Comment by dustymabe** [comment]
> I found https://docs.codacy.com/faq/code-analysis/how-to-skip-an-analysis/ which seems like a way to work around this.

Thanks for the link. I wonder if that only skip it for this run and then new PRs will hit the problem? I'll dig in to see if I can figure it out.

**Comment by dustymabe** [comment]
OK. I just decided to comply rather than waste time trying to figure this out. See the last commit.

**Review by ravanelli — APPROVED** [review]
LGTM, thanks for working in it!

File: jobs/build-arch.Jenkinsfile [hunk 1]
```diff
@@ -141,9 +141,11 @@ lock(resource: "build-${params.STREAM}-${basearch}") {
         // We set the session to time out after 5h. This essentially
         // performs garbage collection on the remote if we fail to clean up.
         pipeutils.withPodmanRemoteArchBuilder(arch: basearch) {
-        def session = shwrapCapture("""
-        cosa remote-session create --image ${cosa_img} --expiration ${timeout_mins}m --workdir ${env.WORKSPACE}
-        """)
+        session = pipeutils.createCosaRemoteSession(
```
Comment (jlebon) [inline jobs/build-arch.Jenkinsfile]: def

File: jobs/debug-pod.Jenkinsfile [hunk 2]
```diff
@@ -72,9 +72,11 @@ cosaPod(cpu: "${ncpus}",
         // We set the session to time out after 4h. This essentially
         // performs garbage collection on the remote if we fail to clean up.
         pipeutils.withPodmanRemoteArchBuilder(arch: params.ARCH) {
-        def session = shwrapCapture("""
-        cosa remote-session create --image ${cosa_img} --expiration ${params.TIMEOUT}h --workdir ${env.WORKSPACE}
-        """)
+        session = pipeutils.createCosaRemoteSession(
```
Comment (jlebon) [inline jobs/debug-pod.Jenkinsfile]: def

File: utils.groovy [hunk 3]
```diff
@@ -217,6 +217,25 @@ def bump_builds_json(stream, buildid, arch, s3_stream_dir, acl) {
...
 
+// Create a COSA remote session, which is usually used to
+// build on a different architecture.
+//
+// Available parameters:
+//    expiration: string that represents a golang duration for how
+//                long the container should last (i.e. 4h, 30m)
+//    image: string that represents the container image to pull
+//    workdir: string that represents the in container working directory
+def createCosaRemoteSession(params = [:]) {
```
Comment (jlebon) [inline utils.groovy]: Aside: this (and a bunch of other functions in this file) would be a good candidate to put in coreos-ci-lib.

File: utils.groovy [hunk 4]
```diff
@@ -123,8 +123,22 @@ def get_source_config_ref_for_stream(pipecfg, stream) {
     }
 }
 
-def get_use_osbuild_for_stream(pipecfg, stream) {
-    return pipecfg.streams[stream].use_osbuild ? '1' : ''
+def get_env_vars_for_stream(pipecfg, stream) {
+    env = [:]
```
Comment (jlebon) [inline utils.groovy:127]: def

File: utils.groovy [hunk 5]
```diff
@@ -123,8 +123,22 @@ def get_source_config_ref_for_stream(pipecfg, stream) {
...
+    // Grab the globally configured env vars first
+    if (pipecfg.env) {
+        for (x in pipecfg.env) {
+            env[x.key] = x.value
+        }
```
Comment (jlebon) [inline utils.groovy:132]: Isn't that just

```groovy
env.putAll(pipecfg.env)
```

?

Similarly below.

File: utils.groovy [hunk 6]
```diff
@@ -221,17 +235,23 @@ def bump_builds_json(stream, buildid, arch, s3_stream_dir, acl) {
...
 //    workdir: string that represents the in container working directory
 def createCosaRemoteSession(params = [:]) {
+    def env = params['env']
     def expiration = params['expiration']
     def image = params['image']
     def workdir = params['workdir']
+    def envArgs = ""
+    for (x in env) {
+        envArgs += "--env ${x.key}=${x.value} "
+    }
```
Comment (jlebon) [inline utils.groovy:251]: What we do elsewhere is e.g.

```
envArgs = env.collect { k, v -> "--env ${k}=${v}" }
```

and then `${envArgs.join(" ")}` below.

File: utils.groovy [hunk 7]
```diff
@@ -221,17 +235,23 @@ def bump_builds_json(stream, buildid, arch, s3_stream_dir, acl) {
...
 //                long the container should last (i.e. 4h, 30m)
 //    image: string that represents the container image to pull
 //    workdir: string that represents the in container working directory
 def createCosaRemoteSession(params = [:]) {
+    def env = params['env']
     def expiration = params['expiration']
     def image = params['image']
     def workdir = params['workdir']
+    def envArgs = ""
+    for (x in env) {
```
Comment (jlebon) [inline utils.groovy:249]: This assumes `env` is non-null, no?

I think we need something like `def env = params['env'] ?: [:]` above, like we do elsewhere.

File: utils.groovy (hunk 5 above)
Comment (dustymabe) [reply utils.groovy:132]: maybe :) - I use groovy about once every 3-6 months so I always end up dropping back to languange non-specific primitives like a for loop.

I'll have to test to make sure `putAll` gives us what we want here.

File: utils.groovy (hunk 6 above)
Comment (dustymabe) [reply utils.groovy:251]: will do

File: utils.groovy (hunk 3 above)
Comment (dustymabe) [reply utils.groovy]: yep - for another day

File: utils.groovy [hunk 8]
```diff
@@ -217,6 +227,28 @@ def bump_builds_json(stream, buildid, arch, s3_stream_dir, acl) {
...
+// build on a different architecture.
+//
+// Available parameters:
+//    env: (optional) map of key/val pairs of environment variables to set
+//    expiration: string that represents a golang duration for how
+//                long the container should last (i.e. 4h, 30m)
+//    image: string that represents the container image to pull
+//    workdir: string that represents the in container working directory
+def makeCosaRemoteSession(params = [:]) {
+    def env = params['env'] ?: [:]
```
Comment (jlebon) [inline utils.groovy:240]: Though note this is shadowing the builtin `env` var. I don't think it matters in this case, but would be cleaner to not shadow.

**Review by jlebon — APPROVED** [review]

File: utils.groovy (hunk 8 above)
Comment (dustymabe) [reply utils.groovy]: fixed

**Review by jlebon — APPROVED** [review]

File: utils.groovy (hunk 8 above)
Comment (dustymabe) [reply utils.groovy]: reworked things to not shadow the `env` var

**Pull request merged**